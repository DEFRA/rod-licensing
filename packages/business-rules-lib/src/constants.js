/**
 * The maximum days in advance of the purchase date that the licence may start
 * @type {number}
 */
export const ADVANCED_PURCHASE_MAX_DAYS = 30

/**
 * The  maximum number of permissions in a (multi-buy) transaction
 * @type {number}
 */
export const MAX_PERMISSIONS_PER_TRANSACTION = 50

/**
 * Allowed status values for the payment journal
 */
export const PAYMENT_JOURNAL_STATUS_CODES = {
  InProgress: 'In Progress',
  Cancelled: 'Cancelled',
  Failed: 'Failed',
  Expired: 'Expired',
  Completed: 'Completed'
}

/**
 * Payment status returned by the GOV.UK pay API
 */
export const PAYMENT_STATUS = {
  Success: 'success',
  Failure: 'failed'
}

/**
 * Failure states returned by the GOV.UK pay API
 * https://docs.payments.service.gov.uk/api_reference/#errors-caused-by-payment-statuses
 */
export const GOVUK_PAY_ERROR_STATUS_CODES = {
  REJECTED: 'P0010',
  EXPIRED: 'P0020',
  USER_CANCELLED: 'P0030',
  ERROR: 'P0050',
  NOT_FOUND: 'P0200'
}

/**
 * Constants used in finalization
 * @type {{telesales: string, govPay: string}}
 */
export const TRANSACTION_SOURCE = { govPay: 'Gov Pay', telesales: 'Telesales' }
export const PAYMENT_TYPE = { debit: 'Debit card' }

/**
 * Data sources associated with POCL transactions
 */
export const POCL_DATA_SOURCE = 'Post Office Sales'
export const DDE_DATA_SOURCE = 'DDE File'
export const POCL_TRANSACTION_SOURCES = [POCL_DATA_SOURCE, DDE_DATA_SOURCE]

/**
 * Describes the lifetime of the renewal link.
 */
export const RENEW_BEFORE_DAYS = 60 // The number of days before licence expiry the link is active
export const RENEW_AFTER_DAYS = 60 // The number of days after licence expiry the link is active

/**
 * The number of minutes after payment that a licence will start
 * @type {number}
 */
export const START_AFTER_PAYMENT_MINUTES = 30

/**
 * Timezone of the service
 */
export const SERVICE_LOCAL_TIME = 'Europe/London'

/**
 * Date for switching paper fulfilment provider
 */
export const FULFILMENT_SWITCHOVER_DATE = '2024-05-30T23:00:00.000Z'
