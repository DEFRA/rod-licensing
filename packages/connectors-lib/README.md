# Rod Licensing - Connectors Lib

Provides connectivity to the resources/infrastructure used in the rod licensing services

# Environment variables

| name                                | description                                                        | required | default         | valid                                                                                          | notes                                                                       |
| ----------------------------------- | ------------------------------------------------------------------ | :------: | --------------- | ---------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------- |
| AWS_REGION                          | The AWS region to use                                              |   yes    |                 | See [AWS Regions](https://docs.aws.amazon.com/general/latest/gr/rande.html#regional-endpoints) |                                                                             |
| AWS_ACCESS_KEY_ID                   | AWS Access Key (required if running outside of AWS or without IAM) |  maybe   |                 | development, test, production                                                                  | Set to "local" when running with local infrastructure                       |
| AWS_SECRET_ACCESS_KEY               | AWS Secret Key (required if running outside of AWS or without IAM) |  maybe   |                 | development, test, production                                                                  | Set to "local" when running with local infrastructure                       |
| AWS_S3_ENDPOINT                     | Custom S3 Endpoint                                                 |    no    | Region specific |                                                                                                | Used to override the S3 service endpoint for local development              |
| AWS_SQS_ENDPOINT                    | Custom SQS Endpoint                                                |    no    | Region specific |                                                                                                | Used to override the SQS service endpoint for local development             |
| AWS_DYNAMODB_ENDPOINT               | Custom DynamoDB Endpoint                                           |    no    | Region specific |                                                                                                | Used to override the DynamoDB service endpoint for local development        |
| AWS_SECRETS_MANAGER_ENDPOINT        | Custom Secrets Manager Endpoint                                    |    no    | Region specific |                                                                                                | Used to override the Secrets Manager service endpoint for local development |
| AWS_NODEJS_CONNECTION_REUSE_ENABLED | Enable connection keep-alive in the AWS connectors                 |    no    | Disabed         | 1 or 0                                                                                         | Recommended to enable connection reuse / keep-alive                         |
| SALES_API_URL                       | URL for the sales API                                              |    no    |                 |                                                                                                |                                                                             |
| SALES_API_TIMEOUT_MS                | Request timeout for the requests to the sales API                  |    no    | 20000 (20s)     |                                                                                                |                                                                             |
| GOV_PAY_API_URL                     | The GOV.UK Pay API base url                                        |   yes    |                 |                                                                                                |                                                                             |
| GOV_PAY_APIKEY                      | GOV pay access identifier                                          |   yes    |                 |                                                                                                |                                                                             |
| GOV_PAY_REQUEST_TIMEOUT_MS          | Timeout in milliseconds for API requests                           |    no    | 10000           |                                                                                                |                                                                             |
| GOV_PAY_RCP_API_URL                 | The GOV.UK Pay API url for agreements                              |   yes    |                 |

# Prerequisites

See [main project documentation](../../README.md).
