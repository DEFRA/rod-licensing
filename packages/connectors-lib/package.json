{
  "name": "@defra-fish/connectors-lib",
  "version": "1.22.0-rc.25",
  "description": "Shared connectors",
  "type": "module",
  "engines": {
    "node": ">=14"
  },
  "keywords": [
    "rod",
    "licensing",
    "sales",
    "api"
  ],
  "author": "DEFRA",
  "license": "SEE LICENSE IN LICENSE",
  "main": "src/connectors.js",
  "directories": {
    "lib": "src",
    "test": "src/__tests__"
  },
  "files": [
    "src"
  ],
  "repository": {
    "type": "git",
    "url": "git://github.com/DEFRA/rod-licensing.git",
    "directory": "packages/connectors-lib"
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "test": "echo \"Error: run tests from root\" && exit 1"
  },
  "dependencies": {
    "@airbrake/node": "^1.4.2",
    "aws-sdk": "^2.922.0",
    "debug": "^4.3.1",
    "ioredis": "^4.27.5",
    "node-fetch": "^2.6.1",
    "redlock": "^4.2.0"
  }
}
