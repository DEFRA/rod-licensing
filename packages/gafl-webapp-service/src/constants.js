/**
 * System constants and defaults
 */
export const ADDRESS_LOOKUP_SERVICE = { lang: 'EN', dataset: 'DPA' }
export const ADDRESS_LOOKUP_TIMEOUT_MS_DEFAULT = 10000
export const ATTRIBUTION_REDIRECT_DEFAULT = '/'
export const SESSION_TTL_MS_DEFAULT = 3 * 60 * 60 * 1000
export const PORT_DEFAULT = 3000
export const REDIS_PORT_DEFAULT = 6379
export const PAGE_STATE = { completed: true, error: false }
export const SESSION_COOKIE_NAME_DEFAULT = 'sid'
export const CSRF_TOKEN_COOKIE_NAME_DEFAULT = 'rlsctkn'
export const FEEDBACK_URI_DEFAULT = '#'
export const CHANNEL_DEFAULT = 'websales'
export const SERVICE_PAGE_DEFAULT = 'https://www.gov.uk/fishing-licences/buy-a-fishing-licence'

export const COMPLETION_STATUS = {
  agreed: 'agreed',
  posted: 'posted',
  paymentCreated: 'payment-created',
  paymentCancelled: 'payment-cancelled',
  paymentFailed: 'payment-failed',
  paymentCompleted: 'payment-completed',
  finalised: 'finalised',
  completed: 'completed'
}

export const GOVPAYFAIL = {
  prePaymentRetry: { step: 'pre-payment' },
  postPaymentRetry: { step: 'post-payment' }
}

export const UTM = {
  CAMPAIGN: 'utm_campaign',
  MEDIUM: 'utm_medium',
  SOURCE: 'utm_source',
  TERM: 'utm_term',
  CONTENT: 'utm_content'
}

export const dateFormats = ['YYYY-MM-DD', 'YY-MM-DD', 'YYYY-M-DD', 'YY-M-DD', 'YYYY-MM-D', 'YY-MM-D', 'YYYY-M-D', 'YY-M-D']

export const CommonResults = {
  SUMMARY: 'summary',
  OK: 'ok',
  NO: 'no',
  YES: 'yes'
}

export const allowsPhysicalLicence = {
  YES: 'yes',
  NO: 'no'
}

// If the user has seen the summary page these are set in the status
export const CONTACT_SUMMARY_SEEN = 'contact-summary'
export const LICENCE_SUMMARY_SEEN = 'licence-summary'

// These cookies are used by the load balancer
export const ALB_COOKIE_NAME = 'AWSALBTG'
export const ALBCORS_COOKIE_NAME = 'AWSALBTGCORS'
