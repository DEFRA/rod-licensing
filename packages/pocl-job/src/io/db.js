import config from '../config.js'
import { docClient } from '../../../connectors-lib/src/aws.js'
import { GetCommand, UpdateCommand } from '@aws-sdk/lib-dynamodb'

/**
 * Update the POCL file staging table to add or update the entry for the provided filename
 *
 * @param {string} filename the primary key of the record in the file staging table to add/update
 * @param {Object} entries to update the table with
 * @returns {Promise<void>}
 */
export const updateFileStagingTable = async ({ filename, ...entries }) => {
  const updateExpression = docClient.createUpdateExpression({
    expires: Math.floor(Date.now() / 1000) + config.db.stagingTtlDelta,
    ...entries
  })
  await docClient.send(
    new UpdateCommand({
      TableName: config.db.fileStagingTable,
      Key: { filename },
      ...updateExpression
    })
  )
}

/**
 * Retrieve all file records for the specified stages.  If stages are not provided then all records are returned
 *
 * @param {string} stages the stage names to filter the result-set on. Omit to retrieve all records regardless of the stage.
 * @returns {Promise<[DocumentClient.AttributeMap]>}
 */
export const getFileRecords = async (...stages) => {
  const stageValues = stages.reduce((acc, s, i) => ({ ...acc, [`:stage${i}`]: s }), {})
  return docClient.scanAllPromise({
    TableName: config.db.fileStagingTable,
    ...(stages.length && { FilterExpression: `stage IN (${Object.keys(stageValues)})` }),
    ExpressionAttributeValues: stageValues,
    ConsistentRead: true
  })
}

/**
 * Retrieve an individual file record by the specified filename
 *
 * @param filename the name of the POCL file to retrieve a record for
 * @returns {DocumentClient.AttributeMap}
 */
export const getFileRecord = async filename => {
  const result = await docClient.send(
    new GetCommand({
      TableName: config.db.fileStagingTable,
      Key: { filename },
      ConsistentRead: true
    })
  )
  return result.Item
}

/**
 * Update the POCL record staging table to add entries for each of the provided records
 *
 * @param {string} filename the filename of a POCL file to which the records relate
 * @param {Array<Object>} records to update the table with
 * @returns {Promise<void>}
 */
export const updateRecordStagingTable = async (filename, records) => {
  if (records.length) {
    const params = {
      RequestItems: {
        [config.db.recordStagingTable]: records.map(record => ({
          PutRequest: { Item: { filename, expires: Math.floor(Date.now() / 1000) + config.db.stagingTtlDelta, ...record } }
        }))
      }
    }
    await docClient.batchWriteAllPromise(params)
  }
}

/**
 * Retrieve processed records from the record staging table for the provided filename and stage filters
 *
 * @param {string} filename the filename of a POCL file for which the records should be retrieved
 * @param {string} stages the stage names to filter the result-set on. Omit to retrieve all records regardless of the stage.
 * @returns {Promise<[DocumentClient.AttributeMap]>}
 */
export const getProcessedRecords = async (filename, ...stages) => {
  const stageValues = stages.reduce((acc, s, i) => ({ ...acc, [`:stage${i}`]: s }), {})
  return docClient.queryAllPromise({
    TableName: config.db.recordStagingTable,
    KeyConditionExpression: 'filename = :filename',
    // ...(stages.length && { FilterExpression: `stage IN (${Object.keys(stageValues).join(',')})` }),
    ...(stages.length && { FilterExpression: `stage IN (${Object.keys(stageValues)})` }),
    ExpressionAttributeValues: { ':filename': filename, ...stageValues },
    ConsistentRead: true
  })
}
