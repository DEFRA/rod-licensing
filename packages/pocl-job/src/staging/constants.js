export const MAX_CREATE_TRANSACTION_BATCH_SIZE = 25
export const MAX_FINALISE_TRANSACTION_BATCH_SIZE = 10

export const POST_OFFICE_DATASOURCE = 'Post Office Sales'
export const DIRECT_DEBIT_DATASOURCE = 'DDE File'
export const DIRECT_DEBIT_PAYMENTSOURCE = 'Direct Debit'
export const POSTAL_ORDER_DATASOURCE = 'Postal Order Sales'
export const POSTAL_ORDER_PAYMENTSOURCE = 'Postal Order'

export const FILE_STAGE = {
  Pending: 'Received and Pending',
  Staging: 'Staging Transactions',
  Finalising: 'Finalising Transactions',
  Completed: 'Completed'
}

export const DYNAMICS_IMPORT_STAGE = {
  Pending: 'Received and Pending',
  InProgress: 'In Progress',
  Processed: 'Processed',
  ProcessedWithWarnings: 'Processed with Warnings',
  Failed: 'Failed',
  UnderReview: 'Under Review',
  Closed: 'Closed',
  isAlreadyProcessed: status => status !== DYNAMICS_IMPORT_STAGE.Pending && status !== DYNAMICS_IMPORT_STAGE.InProgress
}

export const RECORD_STAGE = {
  TransactionCreated: 'Transaction Created',
  TransactionCreationFailed: 'Transaction Creation Failed',
  TransactionFinalised: 'Transaction Finalised',
  TransactionFinalisationFailed: 'Transaction Finalisation Failed'
}
