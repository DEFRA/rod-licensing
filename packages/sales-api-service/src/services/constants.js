const CONCESSION = {
  SENIOR: 'Senior',
  JUNIOR: 'Junior',
  DISABLED: 'Disabled'
}

const CONCESSION_PROOF = {
  none: 'No Proof',
  blueBadge: 'Blue Badge',
  NI: 'National Insurance Number'
}

export { CONCESSION, CONCESSION_PROOF }
