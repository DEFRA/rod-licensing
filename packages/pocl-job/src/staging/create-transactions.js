import { salesApi } from '@defra-fish/connectors-lib'
import Path from 'path'
import { MAX_CREATE_TRANSACTION_BATCH_SIZE, RECORD_STAGE } from './constants.js'
import { transform } from '../transform/pocl-transform-stream.js'
import { getProcessedRecords, updateRecordStagingTable } from '../io/db.js'
import db from 'debug'
const debug = db('pocl:staging')

/**
 * Create the transactions in the Sales API by processing the file at the given path
 *
 * @param xmlFilePath the full file path to the POCL XML file
 * @returns {Promise<{failed: number, succeeded: number}>}
 */
export const createTransactions = async xmlFilePath => {
  const filename = Path.basename(xmlFilePath)
  const state = await getInitialState(filename)

  await transform(xmlFilePath, async function * (source) {
    for await (const data of source) {
      if (!state.processedIds.has(data.id)) {
        state.processedIds.add(data.id)
        state.buffer.push(data)
        if (state.buffer.length === MAX_CREATE_TRANSACTION_BATCH_SIZE) {
          await createTransactionsInSalesApi(filename, state)
        }
      }
    }
  })
  // Process any remaining content of the buffer
  await createTransactionsInSalesApi(filename, state)
  return { succeeded: state.succeeded, failed: state.failed }
}

/**
 * Get the inital state (supports resuming an interrupted process)
 *
 * @typedef TransactionCreationState
 * @property {Array<Object>} buffer the current buffer being processed
 * @property {number} succeeded the number of records which were successfully created in the Sales Api
 * @property {number} failed the number of records which failed to be created in the Sales Api
 * @property {Set<string>} processedIds the ID's of records that were successfully processed
 *
 * @param filename the name of the file being processed (not including path)
 * @returns {Promise<TransactionCreationState>}
 */
const getInitialState = async filename => {
  const processedRecords = await getProcessedRecords(filename)
  processedRecords.length && debug('File partially processed, resuming from record %s', processedRecords.length)
  return processedRecords.reduce(
    (acc, record) => {
      if (record.stage === RECORD_STAGE.TransactionCreationFailed) {
        acc.failed++
      } else {
        acc.succeeded++
      }
      acc.processedIds.add(record.id)
      return acc
    },
    { buffer: [], succeeded: 0, failed: 0, processedIds: new Set() }
  )
}

/**
 * Create the given records in the Sales API, updating properties as appropriate based on whether the record was successfully created or not
 *
 * @param {string} filename the name of the file being processed
 * @param {TransactionCreationState} state the current state of the process
 * @returns {Promise<void>}
 */
const createTransactionsInSalesApi = async (filename, state) => {
  if (state.buffer.length) {
    const createResults = await salesApi.createTransactions(state.buffer.map(item => item.createTransactionPayload))

    const succeeded = []
    const failed = []
    state.buffer.forEach((record, idx) => {
      const result = createResults[idx]
      ;(result.statusCode === 201 ? succeeded : failed).push({ record, result })
    })

    await processSucceeded(filename, succeeded)
    await processFailed(filename, failed)

    state.succeeded += succeeded.length
    state.failed += failed.length
    state.buffer = []
  }
}

/**
 * Update the status of records which were successfully created in the sales API
 *
 * @param {string} filename the name of the file being processed
 * @param {Array<Object>} succeeded An array of objects.  Each object shall contain a record key and a result key
 * @returns {Promise<void>}
 */
const processSucceeded = async (filename, succeeded) => {
  const recordUpdates = succeeded.map(({ record, result }) => {
    record.stage = RECORD_STAGE.TransactionCreated
    record.createTransactionId = result.response.id
    debug('Successfully created transaction for record: %o', record)
    return record
  })
  await updateRecordStagingTable(filename, recordUpdates)
}

/**
 * Update the status of records which failed to be created in the sales API
 *
 * @param {string} filename the name of the file being processed
 * @param {Array<Object>} failed An array of objects.  Each object shall contain a record key and a result key
 * @returns {Promise<void>}
 */
const processFailed = async (filename, failed) => {
  // For record data errors, update the record status to be failed and insert a staging exception into Dynamics
  // We don't need to deal with individual 5xx system errors here as we use the batch endpoint of the sales API
  for (const { record, result } of failed) {
    record.stage = RECORD_STAGE.TransactionCreationFailed
    record.createTransactionError = result
    debug('Failed to create transaction for record: %o', record)
    await salesApi.createStagingException({
      transactionFileException: {
        name: `${filename}: FAILED-CREATE-${record.id}`,
        description: JSON.stringify(result, null, 2),
        json: JSON.stringify(record, null, 2),
        notes: 'Failed to create the transaction in the Sales API',
        type: 'Failure',
        transactionFile: filename
      },
      record
    })
  }

  const recordUpdates = failed.map(({ record }) => record)
  await updateRecordStagingTable(filename, recordUpdates)
}
