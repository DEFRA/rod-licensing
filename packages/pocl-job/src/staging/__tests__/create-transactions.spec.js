import Project from '../../project.cjs'
import { createTransactions } from '../create-transactions.js'
import { RECORD_STAGE, MAX_CREATE_TRANSACTION_BATCH_SIZE } from '../constants.js'
import * as db from '../../io/db.js'
import { v4 as uuidv4 } from 'uuid'
import { salesApi } from '@defra-fish/connectors-lib'

jest.mock('@defra-fish/connectors-lib', () => ({
  salesApi: {
    ...Object.keys(jest.requireActual('@defra-fish/connectors-lib').salesApi).reduce((acc, k) => ({ ...acc, [k]: jest.fn() }), {}),
    concessions: { find: jest.fn(() => ({ id: 'test' })) },
    permits: { find: jest.fn(() => ({ id: 'test', isForFulfilment: true })) }
  }
}))
jest.mock('../../io/db.js', () => ({
  updateRecordStagingTable: jest.fn(),
  getProcessedRecords: jest.fn(() => [])
}))

describe('create-transactions', () => {
  beforeAll(() => {
    process.env.POCL_FILE_STAGING_TABLE = 'TestFileTable'
    process.env.POCL_RECORD_STAGING_TABLE = 'TestRecordTable'
  })
  beforeEach(jest.clearAllMocks)

  it('stages the 2 record test file (under batch-size boundary)', async () => {
    salesApi.createTransactions.mockReturnValue(generateApiSuccessResponse(2))
    await createTransactions(`${Project.root}/src/__mocks__/test-2-records.xml`)
    expectCreateTransactionCalls(2)
    expect(db.updateRecordStagingTable.mock.calls).toHaveLength(2)
    // First call to update records which were successfully created
    expect(db.updateRecordStagingTable.mock.calls[0][1]).toHaveLength(2)
    expect(db.updateRecordStagingTable.mock.calls[0][1]).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          createTransactionId: expect.any(String),
          stage: RECORD_STAGE.TransactionCreated
        })
      ])
    )
    // Second call to update records which failed (there shouldn't be any for this test)
    expect(db.updateRecordStagingTable.mock.calls[1][1]).toHaveLength(0)
  })

  it('stages the 25 record test file (on batch-size boundary)', async () => {
    salesApi.createTransactions.mockReturnValue(generateApiSuccessResponse(MAX_CREATE_TRANSACTION_BATCH_SIZE))
    await createTransactions(`${Project.root}/src/__mocks__/test-25-records.xml`)
    expectCreateTransactionCalls(MAX_CREATE_TRANSACTION_BATCH_SIZE)
    // First call to update records which were successfully created
    expect(db.updateRecordStagingTable.mock.calls).toHaveLength(2)
    expect(db.updateRecordStagingTable.mock.calls[0][1]).toHaveLength(MAX_CREATE_TRANSACTION_BATCH_SIZE)
    expect(db.updateRecordStagingTable.mock.calls[0][1]).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          createTransactionId: expect.any(String),
          stage: RECORD_STAGE.TransactionCreated
        })
      ])
    )
    // Second call to update records which failed (there shouldn't be any for this test)
    expect(db.updateRecordStagingTable.mock.calls[1][1]).toHaveLength(0)
  })

  it('stages the 30 record test file (over batch-size boundary)', async () => {
    salesApi.createTransactions.mockReturnValue(generateApiSuccessResponse(30))
    await createTransactions(`${Project.root}/src/__mocks__/test-30-records.xml`)
    expectCreateTransactionCalls(MAX_CREATE_TRANSACTION_BATCH_SIZE, 5)
    expect(db.updateRecordStagingTable.mock.calls).toHaveLength(4)
    // First call of first batch to update records which were successfully created
    expect(db.updateRecordStagingTable.mock.calls[0][1]).toHaveLength(MAX_CREATE_TRANSACTION_BATCH_SIZE)
    expect(db.updateRecordStagingTable.mock.calls[0][1]).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          createTransactionId: expect.any(String),
          stage: RECORD_STAGE.TransactionCreated
        })
      ])
    )
    // Second call of first batch to update records which failed (there shouldn't be any for this test)
    expect(db.updateRecordStagingTable.mock.calls[1][1]).toHaveLength(0)

    // First call of second batch to update records which were successfully created
    expect(db.updateRecordStagingTable.mock.calls[2][1]).toHaveLength(5)
    expect(db.updateRecordStagingTable.mock.calls[2][1]).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          createTransactionId: expect.any(String),
          stage: RECORD_STAGE.TransactionCreated
        })
      ])
    )
    // Second call of second batch to update records which failed (there shouldn't be any for this test)
    expect(db.updateRecordStagingTable.mock.calls[3][1]).toHaveLength(0)
  })

  it('resumes from the second record of the 2 record test file if the first record has already been processed successfully', async () => {
    salesApi.createTransactions.mockReturnValue(generateApiSuccessResponse(1))
    db.getProcessedRecords.mockReturnValueOnce([{ id: 'SERIAL 1', stage: RECORD_STAGE.TransactionCreated }])
    await createTransactions(`${Project.root}/src/__mocks__/test-2-records.xml`)
    expectCreateTransactionCalls(1)
    expect(db.updateRecordStagingTable.mock.calls).toHaveLength(2)
    // First call to update records which were successfully created
    expect(db.updateRecordStagingTable.mock.calls[0][1]).toHaveLength(1)
    expect(db.updateRecordStagingTable.mock.calls[0][1]).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          createTransactionId: expect.any(String),
          stage: RECORD_STAGE.TransactionCreated
        })
      ])
    )
    // Second call to update records which failed (there shouldn't be any for this test)
    expect(db.updateRecordStagingTable.mock.calls[1][1]).toHaveLength(0)
  })

  it('resumes from the second record of the 2 record test file if the first record has already been processed and failed', async () => {
    salesApi.createTransactions.mockReturnValue(generateApiSuccessResponse(1))
    db.getProcessedRecords.mockReturnValueOnce([{ id: 'SERIAL 1', stage: RECORD_STAGE.TransactionCreationFailed }])
    await createTransactions(`${Project.root}/src/__mocks__/test-2-records.xml`)
    expectCreateTransactionCalls(1)
    expect(db.updateRecordStagingTable.mock.calls).toHaveLength(2)
    // First call to update records which were successfully created
    expect(db.updateRecordStagingTable.mock.calls[0][1]).toHaveLength(1)
    expect(db.updateRecordStagingTable.mock.calls[0][1]).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          createTransactionId: expect.any(String),
          stage: RECORD_STAGE.TransactionCreated
        })
      ])
    )
    // Second call to update records which failed (there shouldn't be any for this test)
    expect(db.updateRecordStagingTable.mock.calls[1][1]).toHaveLength(0)
  })

  it('handles exceptions', async () => {
    salesApi.createTransactions.mockReturnValue(generateApiResponses(201, 422))
    const fakeApiError = { statusCode: 422, error: 'Fake error', message: 'Fake error message' }
    await createTransactions(`${Project.root}/src/__mocks__/test-2-records.xml`)
    expectCreateTransactionCalls(2)
    expect(db.updateRecordStagingTable.mock.calls).toHaveLength(2)
    expect(db.updateRecordStagingTable.mock.calls[0][1]).toHaveLength(1)
    expect(db.updateRecordStagingTable.mock.calls[0][1][0]).toEqual(
      expect.objectContaining({
        createTransactionId: expect.any(String),
        stage: RECORD_STAGE.TransactionCreated
      })
    )
    expect(db.updateRecordStagingTable.mock.calls[1][1][0]).toEqual(
      expect.objectContaining({
        createTransactionError: fakeApiError,
        createTransactionPayload: expect.objectContaining({}),
        stage: RECORD_STAGE.TransactionCreationFailed
      })
    )
    expect(salesApi.createStagingException).toHaveBeenCalledWith(
      expect.objectContaining({
        transactionFileException: {
          name: 'test-2-records.xml: FAILED-CREATE-SERIAL 2',
          description: JSON.stringify(fakeApiError, null, 2),
          json: expect.any(String),
          transactionFile: 'test-2-records.xml',
          type: 'Failure',
          notes: 'Failed to create the transaction in the Sales API'
        }
      })
    )
  })
  it('adds record to staging exception', async () => {
    salesApi.createTransactions.mockReturnValue(generateApiResponses(201, 422))
    await createTransactions(`${Project.root}/src/__mocks__/test-2-records.xml`)
    const {
      calls: [[{ record }]]
    } = salesApi.createStagingException.mock
    expect(record).toMatchSnapshot()
  })
})

const expectCreateTransactionCalls = (...arrayLengths) => {
  expect(salesApi.createTransactions).toHaveBeenCalledTimes(arrayLengths.length)
  for (let i = 0; i < arrayLengths.length; i++) {
    const argumentPassedToCreateTransactions = salesApi.createTransactions.mock.calls[i][0]
    expect(argumentPassedToCreateTransactions).toBeInstanceOf(Array)
    expect(argumentPassedToCreateTransactions).toHaveLength(arrayLengths[i])
  }
}

const generateApiSuccessResponse = count => {
  const successResponses = Array(count).fill(201)
  return generateApiResponses(...successResponses)
}
const generateApiResponses = (...responseCodes) => {
  const response = []
  for (let i = 0; i < responseCodes.length; i++) {
    const isSuccess = Math.floor(responseCodes[i] / 100) === 2
    response.push({
      statusCode: responseCodes[i],
      ...(isSuccess ? { response: { id: uuidv4() } } : { error: 'Fake error', message: 'Fake error message' })
    })
  }
  return response
}
