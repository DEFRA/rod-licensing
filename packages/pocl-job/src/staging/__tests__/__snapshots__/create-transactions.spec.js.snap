// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-transactions adds record to staging exception 1`] = `
Object {
  "createTransactionError": Object {
    "error": "Fake error",
    "message": "Fake error message",
    "statusCode": 422,
  },
  "createTransactionPayload": Object {
    "dataSource": "Post Office Sales",
    "permissions": Array [
      Object {
        "issueDate": "2020-07-01T12:00:00.000Z",
        "licensee": Object {
          "birthDate": "1975-08-26",
          "country": "England",
          "email": "cleaver321@email.com",
          "firstName": "Daniel",
          "lastName": "Cleaver",
          "mobilePhone": "07763 722080",
          "postcode": "PL5 2AA",
          "preferredMethodOfConfirmation": "Email",
          "preferredMethodOfNewsletter": "Email",
          "preferredMethodOfReminder": "Email",
          "premises": "65",
          "street": "Groveside Close",
          "town": "Carshalton",
        },
        "permitId": "test",
        "startDate": "2020-07-01T14:00:00.000Z",
      },
    ],
    "serialNumber": "SERIAL 2",
  },
  "finaliseTransactionPayload": Object {
    "payment": Object {
      "amount": 82,
      "channelId": "21995X",
      "method": "Stamps",
      "source": "Post Office Sales",
      "timestamp": "2020-07-01T12:00:00.000Z",
    },
  },
  "id": "SERIAL 2",
  "stage": "Transaction Creation Failed",
}
`;
